main/src/mill/modules/Assembly.scala

```diff
@@ -26,11 +26,19 @@ object Assembly {
   object Rule {
     case class Append(path: String, separator: String = defaultSeparator) extends Rule
 
+    @deprecated("Use mill.modules.Assembly.Rule.AppendByPattern instead", since = "mill 0.10.1")
     object AppendPattern {
       def apply(pattern: String): AppendPattern = AppendPattern(Pattern.compile(pattern))
     }
+    @deprecated("Use mill.modules.Assembly.Rule.AppendByPattern instead", since = "mill 0.10.1")
     case class AppendPattern(pattern: Pattern) extends Rule
 
+    object AppendByPattern {
```

Comment:
By creating a new `case class` you're creating a new problem for the next person that will need another parameter.
Then we will have either 3 case classes or a deprecated `case class` and a class with all the ugly workaround you reverted.
I think Mill should not add any more `case class`es to the codebase unless there is a very very strong reason to do so.

Reply:
For me sounds like YAGNI violation, but you know best. Reverting revert...